### DISCUSSION TITLE
Is there a way to run Slint Live Preview with Software Renderer?

### DISCUSSION THREAD

**Discussion by yanshay** [slint-ui/slint#6257] [Q&A]
I'm working on MCU that use Software Renderer, and to speed development I use the Slint Live Preview.
Then when I flash the code to the device, things don't work or work differently and I need to undo my changes and find alternative ways.

A few examples:
1. Text don't support rotate
2. All kinds of gradient limitations and bugs
3. Fonts appearance and capabilities different in device vs. live preview

To get UI to look good its a lot about small tunings and for every small thing I need to do several cycles of flashing and running a complete app instead of just a quick change. 

So maybe it's already supported and I can switch the live preview to the software renderer? I searched but couldn't find a way to do that.

**Comment by ogoffart** [comment]
Yes.

 - Set the environment variable `SLINT_BACKEND=software` before running your editor so that slint-lsp inherit this env variable.
 - Or, In VSCode, there is a setting slint.lsp-args , in which you can add `--backend=software`, and reload the server, or you can use `--backend=software` as an arg to the slint-lsp in other editors that allow it.


![image](https://github.com/user-attachments/assets/cb1a8a19-d462-4db4-87b5-d4acdc05e5d8)

**Reply by yanshay** [reply]
I tried the VSCode setting option and it doesn't change anything displayed. I still see differences from the device and can still see capabilities that don't work on the device. I also tried setting the environment from where I run VSCode and it's the same.

Maybe there are differences between when the software-renderer runs on the device and in the live-preview? These are the differences I see: Different font sizes and text vertical positioning, support for changing font weights dynamically that don't work on the device (probably because it didn't load that font weight into the binary at compile time).

Or maybe the setting didn't take effect? Is there a way to know from the live preview that it is in software-renderer mode?

Is there a way to run the live preview from command line and not through the editor?

**Reply by algovoid** [reply]
Hi, 

Regarding your last question,  you can check this : https://github.com/slint-ui/slint/tree/master/tools/viewer

**Reply by yanshay** [reply]
I tried `slint-viewer` with `software` backend. 

While it looks different than the regular mode of `slint-viewer` it is still different from how things really look like on the device and provides capabilities that don't work on the device. In some ways it differs more than the standard viewer from the actual device.

**Reply by ogoffart** [reply]
What are these difference exactly. 
The one i can think of are:
 - The image are being loaded dynamically, especially the SVG are not being pre-rendered to the same size.
 - The fonts are also being loaded synamically instead of being pre-rendered.

These are the difference I think. Anything else?

Maybe another difference could be the 16bit pixel vs 32 bit pixel? 

Adding a "limited" mode to get a preview more close to the device might be required.

**Reply by yanshay** [reply]
What I saw in a brief check:
1. A symbol character that showed on the MCU doesn't show on the slint-viewer with backend=software. That is strange since w/o the backend=software it showed it (incorrect placement, but showing).
2. It includes the font-weights while the MCU didn't. 

> Adding a "limited" mode to get a preview more close to the device might be required.

Totally agreed.
I'll explain why something small like the 2nd point above is important for example: what I did, was to code that when a button is pressed its text gets emphasized using weight. That caused the button to get widened, so I had to spend time on changing the layout of my buttons so it won't happen. And after all that was done, it didn't work on the MCU. 
It's not that the feature I couldn't implement was that important, but I didn't know it won't work and spent time on that.
So now, I only rely on the device for the small details. And every cycle on the device is a couple of minutes of full compile, flash, run and get to the screen, etc. . So fine tuning the UI becomes really time consuming. Every vertical movement of a pixel or two of text to appear really centered is a couple of minutes to validate.
